pkg/operator/ceph/object/topic/controller_test.go

```diff
@@ -0,0 +1,245 @@
+		s.AddKnownTypes(cephv1.SchemeGroupVersion, &cephv1.CephBucketTopic{}, &cephv1.CephBucketTopicList{}, &cephv1.CephCluster{}, &cephv1.CephClusterList{})
+		cl := fake.NewClientBuilder().WithScheme(s).WithRuntimeObjects(objects...).Build()
+
+		r := &ReconcileBucketTopic{client: cl, context: c, clusterInfo: clusterInfo, clusterSpec: &clusterSpec, opManagerContext: ctx}
+		res, err := r.Reconcile(ctx, req)
+		assert.NoError(t, err)
+		assert.True(t, res.Requeue)
+	})
+
+	t.Run("creating a topic", func(t *testing.T) {
+		// Objects to track in the fake client.
+		objects := []runtime.Object{
+			bucketTopic,
+			&cephv1.CephCluster{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      namespace,
+					Namespace: namespace,
+				},
+				Status: cephv1.ClusterStatus{
+					Phase: k8sutil.ReadyStatus,
+					CephStatus: &cephv1.CephStatus{
+						Health: "HEALTH_OK",
+					},
+				},
+			},
+		}
+
+		executor := &exectest.MockExecutor{
+			MockExecuteCommandWithTimeout: func(timeout time.Duration, command string, args ...string) (string, error) {
+				if args[0] == "user" && args[1] == "create" {
+					return userCreateJSON, nil
+				}
+				return "", nil
+			},
+		}
+
+		c := &clusterd.Context{
+			Executor:      executor,
+			RookClientset: rookclient.NewSimpleClientset(),
+			Clientset:     test.New(t, 3),
+		}
+
+		secrets := map[string][]byte{
+			"fsid":         []byte(name),
+			"mon-secret":   []byte("monsecret"),
+			"admin-secret": []byte("adminsecret"),
+		}
+		secret := &v1.Secret{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      "rook-ceph-mon",
+				Namespace: namespace,
+			},
+			Data: secrets,
+			Type: k8sutil.RookType,
+		}
+		_, err := c.Clientset.CoreV1().Secrets(namespace).Create(ctx, secret, metav1.CreateOptions{})
+		assert.NoError(t, err)
+
+		s := scheme.Scheme
+		s.AddKnownTypes(cephv1.SchemeGroupVersion, &cephv1.CephBucketTopic{}, &cephv1.CephBucketTopicList{}, &cephv1.CephCluster{}, &cephv1.CephClusterList{})
+
+		cl := fake.NewClientBuilder().WithScheme(s).WithRuntimeObjects(objects...).Build()
+
+		cephObjectStore := &cephv1.CephObjectStore{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      store,
+				Namespace: namespace,
+			},
+			TypeMeta: metav1.TypeMeta{
+				Kind: "CephObjectStore"},
+			Spec: cephv1.ObjectStoreSpec{
+				Gateway: cephv1.GatewaySpec{
+					Port: int32(80),
+				},
+			},
+		}
+
+		_, err = c.RookClientset.CephV1().CephObjectStores(namespace).Create(ctx, cephObjectStore, metav1.CreateOptions{})
+		assert.NoError(t, err)
+		r := &ReconcileBucketTopic{client: cl, context: c, clusterInfo: clusterInfo, clusterSpec: &clusterSpec, opManagerContext: ctx}
+
+		err = r.client.Get(ctx, types.NamespacedName{Name: name, Namespace: namespace}, bucketTopic)
+		assert.NoError(t, err, bucketTopic)
+
+		// mock the provisioner
+		expectedARN := "arn:aws:sns:" + store + bucketTopic.Name
+		Create = func(p Provisioner, topic *cephv1.CephBucketTopic) (*string, error) {
+			return &expectedARN, nil
+		}
```

Comment:
Making `Create` take `Provisioner` as an argument is good use of dependency injection and is helpful for unit testing. :+1:

This doesn't really mock the provisioner but instead mocks the `Create` function. Since there are no unit tests for `Create`, the test itself may very well be missing any bugs that are present in the implementation of `Create`. I would suggest creating thorough unit tests for `Create`, and then this test can check that `Create` is (a) called when it should be and (b) not called when it shouldn't (previous sub-tests). This means that this test doesn't have to (but it can if you like) verify the functionality of `Create`, a sort of unit-integration kind of test.

Reply:
i plan on testing the provisioner separately. most complexity is in the controller, which is the entity tested here